pkg/operation/common/extensions.go

```diff
@@ -0,0 +1,263 @@
+		}
+
+		if postReadyFunc != nil {
+			if err := postReadyFunc(obj); err != nil {
+				return retry.SevereError(err)
+			}
+		}
+
+		return retry.Ok()
+	}); err != nil {
+		return gardencorev1beta1helper.DetermineError(err, fmt.Sprintf("Error while waiting for %s to become ready: %v", extensionKey(kind, namespace, name), err))
+	}
+	return nil
+}
+
+// DeleteExtensionCR deletes an extension resource.
+func DeleteExtensionCR(
+	ctx context.Context,
+	c client.Client,
+	newObjFunc func() extensionsv1alpha1.Object,
+	namespace string,
+	name string,
+) error {
+	obj := newObjFunc()
+	obj.SetNamespace(namespace)
+	obj.SetName(name)
+
+	if err := ConfirmDeletion(ctx, c, obj); err != nil {
+		return err
+	}
+
+	return client.IgnoreNotFound(c.Delete(ctx, obj, kubernetes.DefaultDeleteOptions...))
+}
+
+// DeleteExtensionCRs lists all extension resources and loops over them. It executes the given <predicateFunc> for each
+// of them, and if it evaluates to true then the resource will be deleted.
+func DeleteExtensionCRs(
+	ctx context.Context,
+	c client.Client,
+	listObj runtime.Object,
+	newObjFunc func() extensionsv1alpha1.Object,
+	namespace string,
+	predicateFunc func(obj extensionsv1alpha1.Object) bool,
+) error {
+	if err := c.List(ctx, listObj, client.InNamespace(namespace)); err != nil {
+		return err
+	}
+
+	fns := make([]flow.TaskFn, 0, meta.LenList(listObj))
+
+	if err := meta.EachListItem(listObj, func(obj runtime.Object) error {
+		o, ok := obj.(extensionsv1alpha1.Object)
+		if !ok {
+			return fmt.Errorf("expected extensionsv1alpha1.Object but got %T", obj)
+		}
+
+		if predicateFunc != nil && predicateFunc(o) {
+			fns = append(fns, func(ctx context.Context) error {
+				return DeleteExtensionCR(ctx, c, newObjFunc, o.GetNamespace(), o.GetName())
+			})
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+// WaitUntilExtensionCRsDeleted lists all extension resources and loops over them. It executes the given <predicateFunc>
+// for each of them, and if it evaluates to true then it waits for the resource to be deleted.
+func WaitUntilExtensionCRsDeleted(
+	ctx context.Context,
+	c client.Client,
+	logger *logrus.Entry,
+	listObj runtime.Object,
+	newObjFunc func() extensionsv1alpha1.Object,
+	kind string,
+	namespace string,
+	interval time.Duration,
+	timeout time.Duration,
+	predicateFunc func(obj extensionsv1alpha1.Object) bool,
+) error {
+	if err := c.List(ctx, listObj, client.InNamespace(namespace)); err != nil {
+		return err
+	}
+
+	fns := make([]flow.TaskFn, 0, meta.LenList(listObj))
+
+	if err := meta.EachListItem(listObj, func(obj runtime.Object) error {
+		o, ok := obj.(extensionsv1alpha1.Object)
+		if !ok {
+			return fmt.Errorf("expected extensionsv1alpha1.Object but got %T", obj)
+		}
+
+		if o.GetDeletionTimestamp() == nil {
+			return nil
+		}
+
+		if predicateFunc != nil && predicateFunc(o) {
+			fns = append(fns, func(ctx context.Context) error {
```

Comment:
I know that you didn't change the logic but I wondered if we can operate on the list directly instead of adding funcs which again wait for a single deletion?

Reply:
Hm, what do you suggest concretely when you saw "operate on the list directly"?

Reply:
Like getting the list --> apply the predicates --> if elements still remain, run the check again. So my proposal is to use the elements from the list instead of retrieving them separately again.

Reply:
Ok, understood, thanks. I tried it, but I ended up having again two places where the `lastError` of the extension object is evaluated. I can introduce helper functions for that, however, as this PR is a preparation for follow-up work that properly evaluates the error codes and I don't know how this will look like, would it be okay to keep it like this for now? I keep it in mind and check whether it can be implemented after the error codes improvements.

Reply:
Yep, sound reasonable. Thanks for your effort checking this.